### PR TITLE
fix: improve sdk loading snippet

### PR DESCRIPTION
**Pull request by saikumarrs**
## PR Description

A bunch of improvements are made in the loading snippet.

- Avoid direct usage of `window.rudderanalytics` and replaced with dynamic usage `window[key` (key is `rudderanalytics`) in the snippet to further fly under the radar of adblockers parsing the loading snippet.
- The snippet considers a case where `window.rudderanalytics` is already defined. Otherwise, defines it as an empty array.
- If the SDK is already loaded, the snippet skips everything.
- If the snippet is included twice, an error is logged onto the console.
- Along with asynchronous script loading, we've also introduced an option to load it in 'defer' mode.
- Most importantly, the stub API methods now check if SDK is already loaded or not and accordingly execute the respective statements.
- Instead of appending the script tags to the existing collection, we're now making sure they are prepended ahead of all the script tags on the webpage.
- A new attribute `loader` with value `RS_JS_SDK` is set to the script tags.
- For the core SDK script, additionally, `rsaWriteKey` attribute is set to the write key of the source. This will enable us to properly identify the script tag on the webpage even if the file name is not a standard one.

Additionally, the default attribute set in the external source loader module is renamed to `loader` to ensure consistency.

## Linear task (optional)

https://linear.app/rudderstack/issue/SDK-1948/handle-api-function-references-for-preloaded-events-in-the-loading

## Cross Browser Tests

Please confirm you have tested for the following browsers:

- [ ] Chrome
- [ ] Firefox
- [ ] IE11

## Sanity Suite

- [ ] All sanity suite test cases pass locally

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

## Summary by CodeRabbit

- **New Features**
  - Updated RudderStack JavaScript SDK snippet to version 3.0.9 across various examples, improving initialization logic and preventing multiple executions.
  - Enhanced script loading mechanisms for better cross-environment compatibility and single SDK loading.

- **Bug Fixes**
  - Added error handling and checks to prevent duplicate inclusion of the RudderStack snippet.

- **Chores**
  - Removed obsolete configuration variables from `.env.example` to streamline environment setup.
  - Improved local development experience with new plugins for HTML templates, live reloading, and local serving based on environment variables.

These updates aim to enhance the reliability and performance of the RudderStack SDK integration in different environments.

### CODE REVIEW

File: packages/analytics-js-common/src/services/ExternalSrcLoader/jsFileLoader.ts [hunk 1]
```diff
@@ -33,12 +33,13 @@ const createScriptElement = (
   scriptElement.src = url;
   scriptElement.id = id;
   scriptElement.async = async;
-  scriptElement.setAttribute('data-append-origin', EXTERNAL_SOURCE_LOAD_ORIGIN);
 
   Object.keys(extraAttributes).forEach(attributeName => {
     scriptElement.setAttribute(attributeName, extraAttributes[attributeName] as string);
   });
 
+  scriptElement.setAttribute('data-loader', EXTERNAL_SOURCE_LOAD_ORIGIN);
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-common/src/services/ExternalSrcLoader/jsFileLoader.ts:41]: Cannot use 'for' due to linting constaints.

File: packages/loading-scripts/src/index.ts [hunk 2]
```diff
@@ -1,106 +1,149 @@
...
+        Object.defineProperty(Object.prototype, '__globalThis_magic__', {
+          get: function get() {
+            return this;
+          },
+          configurable: true,
+        });
+        // @ts-ignore
+        __globalThis_magic__.globalThis = __globalThis_magic__;
+        // @ts-ignore
+        delete Object.prototype.__globalThis_magic__;
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/loading-scripts/src/index.ts:119]: It's fine as it is meant to be a polyfill.

File: packages/loading-scripts/src/index.ts [hunk 3]
```diff
@@ -1,106 +1,149 @@
...
+        // eslint-disable-next-line func-names
+        function () {
+          if (Array.isArray(window[identifier])) {
+            (rudderanalytics as unknown as PreloadedEventCall[]).push(
+              [methodName].concat(Array.prototype.slice.call(arguments) as PreloadedEventCall),
+            );
+          } else {
+            (window[identifier] as any)[methodName]?.apply(window[identifier], arguments);
+          }
+        })(method);
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/loading-scripts/src/index.ts:58]: This code needs to be cross-compatible across all browsers including legacy. Hence, no arrow functions were used.

File: packages/analytics-js/public/index.html [hunk 4]
```diff
@@ -2,147 +2,153 @@
...
+              // },
+              // preConsent: {
+              //   enabled: true,
+              //   storage: {
+              //     strategy: 'session'
+              //   },
+              //   events: {
+              //     delivery: 'buffer'
+              //   }
+              }
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js/public/index.html:139]: They are needed.

File: packages/sanity-suite/public/v3/manualLoadCall/index-cdn.html [hunk 5]
```diff
@@ -21,77 +21,86 @@
...
+                delete Object.prototype.__globalThis_magic__;
+              }
+              window.rudderAnalyticsAddScript("".concat(sdkBaseUrl, "/").concat(window.rudderAnalyticsBuildType, "/").concat(sdkName), "data-rsa-write-key", "__WRITE_KEY__");
+            };
+            if (typeof Promise === "undefined" || typeof globalThis === "undefined") {
+              window.rudderAnalyticsAddScript("https://polyfill-fastly.io/v3/polyfill.min.js?version=3.111.0&features=Symbol%2CPromise&callback=rudderAnalyticsMount");
+            } else {
+              window.rudderAnalyticsMount();
+            }
+          }
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/sanity-suite/public/v3/manualLoadCall/index-cdn.html:101]: No need to change it.

File: packages/analytics-js/public/index.html [hunk 6]
```diff
@@ -2,147 +2,153 @@
...
+                scriptTag.async = true;
+              } else if (deferScript) {
+                scriptTag.defer = true;
+              }
+              if (head) {
+                head.insertBefore(scriptTag, head.firstChild);
+              } else {
+                body.insertBefore(scriptTag, body.firstChild);
+              }
+            };
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js/public/index.html:60]: Can be ignored.

File: examples/angular/sample-app/src/index.html [hunk 7]
```diff
@@ -7,17 +7,37 @@
...
+      var identifier = "rudderanalytics";
+      if (!window[identifier]) {
+        window[identifier] = [];
+      }
+      var rudderanalytics = window[identifier];
+      if (Array.isArray(rudderanalytics)) {
+        if (rudderanalytics.snippetExecuted === true && window.console && console.error) {
+          console.error("RudderStack JavaScript SDK snippet included more than once.");
+        } else {
+          rudderanalytics.snippetExecuted = true;
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply examples/angular/sample-app/src/index.html:22]: It's fine for now.

File: examples/angular/sample-app/src/index.html [hunk 8]
```diff
@@ -7,17 +7,37 @@
...
-      var method = methods[i];
-      window.rudderanalytics[method] = function(methodName) {
-        return function() {
-          window.rudderanalytics.push([ methodName ].concat(Array.prototype.slice.call(arguments)));
-        };
-      }(method);
-    }
+    (function() {
+      "use strict";
+      window.RudderSnippetVersion = "3.0.9";
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply examples/angular/sample-app/src/index.html:12]: It's the correct version.

File: examples/angular/sample-app/src/index.html [hunk 9]
```diff
@@ -7,17 +7,37 @@
...
+              return function() {
+                if (Array.isArray(window[identifier])) {
+                  rudderanalytics.push([ methodName ].concat(Array.prototype.slice.call(arguments)));
+                } else {
+                  var _methodName;
+                  (_methodName = window[identifier][methodName]) === null || _methodName === void 0 || _methodName.apply(window[identifier], arguments);
+                }
+              };
+            }(method);
+          }
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply examples/angular/sample-app/src/index.html:37]: Why do you think it is correct?

File: examples/reactjs/vite/sample-app/index.html [hunk 10]
```diff
@@ -6,17 +6,37 @@
...
+        var identifier = "rudderanalytics";
+        if (!window[identifier]) {
+          window[identifier] = [];
+        }
+        var rudderanalytics = window[identifier];
+        if (Array.isArray(rudderanalytics)) {
+          if (rudderanalytics.snippetExecuted === true && window.console && console.error) {
+            console.error("RudderStack JavaScript SDK snippet included more than once.");
+          } else {
+            rudderanalytics.snippetExecuted = true;
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply examples/reactjs/vite/sample-app/index.html:21]: Already answered in another comment.

File: examples/angular/sample-app/src/index.html (hunk 9 above)
[the remark this answers is not in the document]
Comment (saikumarrs) [reply examples/angular/sample-app/src/index.html:37]: While executing the stub methods, window.rudderanalytics can be an object. Hence, that "if" statement was added for robustness.

File: packages/analytics-js/public/index.html [hunk 11]
```diff
@@ -2,147 +2,153 @@
...
+          if (rudderanalytics.snippetExecuted === true && window.console && console.error) {
+            console.error("RudderStack JavaScript SDK snippet included more than once.");
+          } else {
+            rudderanalytics.snippetExecuted = true;
+            window.rudderAnalyticsBuildType = "legacy";
+            var sdkBaseUrl = "https://cdn.rudderlabs.com/v3";
+            var sdkName = "rsa.min.js";
+            var asyncScript = true;
+            var deferScript = false;
+            var methods = [ "setDefaultInstanceKey", "load", "ready", "page", "track", "identify", "alias", "group", "reset", "setAnonymousId", "startSession", "endSession", "consent" ];
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js/public/index.html]: Done.

File: packages/sanity-suite/public/v3/integrations/index-local.html [hunk 12]
```diff
@@ -21,77 +21,85 @@
...
+              if (scriptLoadingMode === "async") {
+                scriptTag.async = true;
+              } else if (scriptLoadingMode === "defer") {
+                scriptTag.defer = true;
+              }
+              if (head) {
+                head.insertBefore(scriptTag, head.firstChild);
+              } else {
+                body.insertBefore(scriptTag, body.firstChild);
+              }
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/sanity-suite/public/v3/integrations/index-local.html:80]: It'll automatically take care of empty head or body tags.

**Review by MoumitaM — APPROVED** [review]

**Pull request merged**